### ISSUE TITLE
[Research] SaT LoRA-First Validation and Qwen/Soniox Simultaneous-Translation Boundary Recovery

### ISSUE THREAD

**Issue by kapitalismho** [kapitalismho/PuriPuly-heart#49]
## Working Basis — Read This First

- This issue is **not** to be implemented from `dev` or `main`.
- Work must be based on the **separate local experimental branch** already prepared by the user, including its existing experiment code, replay tools, data loaders, and configuration files.
- The experimental branch may exist only locally. Do not assume its name and do not create a new branch.
- Before making changes, inspect the current local checkout and the prepared code. Extend the existing experiment harness instead of rebuilding duplicate infrastructure.
- Do not check out, rebase onto, or cherry-pick from `dev` or `main`, and do not rewrite the work around the current structure of those branches.
- Keep experiment outputs and commits on the user-prepared local experimental branch.

---

## Background

Experiment 11b showed that independently applying a threshold to every raw-partial candidate does not work.

- Deepgram `confirmed source + raw T4 + SaT` A-FIRST remains the strongest baseline.
  - Precision: 90.9%
  - Candidate recall: 76.9%
  - Reference-positive turn coverage: 83.3%
  - Preview p50 / p90: 2,998 ms / 4,845 ms
  - Calls/turn: 0.58
  - Decision flips: 0
- Global profiles that directly trigger from raw-partial candidates did not meet the precision floor for Deepgram or Soniox.
- Qwen's best global profile reached only 38.9% turn coverage.
- Removing the coverage constraint reduced precision to 42.9–55.7% while increasing calls/turn to roughly 3.2.

Current conclusion:

> Keep SaT, but change its role from a threshold classifier that independently decides whether every raw candidate should trigger into a boundary verifier / cut selector operating inside a safe source window.

This issue starts with **official ASR-style SaT LoRA validation**, then applies the winning SaT backend to the Qwen and Soniox follow-up experiments.

---

## Goals

1. Determine whether `sat-3l + ted2020-corrupted LoRA` separates ASR-partial boundaries better than the current base SaT models.
2. For Qwen, select the best boundary retrospectively inside a rollback-free confirmed prefix.
3. For Soniox, use the existing **local VAD + manual finalize** pipeline as a stable source-candidate generator, then let SaT decide whether the text is ready for translation.
4. Compare boundary ranking, persistence, Viterbi / length constraints, and bounded-latency scheduling instead of relying only on a fixed threshold.
5. Allow provider-specific source builders while preserving a shared `TranslationUpdate` contract.

---

# P0 — Official SaT LoRA Re-evaluation (Run First)

## Hypothesis

Compared with a generic SaT model, the official TED ASR-style adapter may separate `WAIT` boundaries from useful translation boundaries more effectively when candidates contain fillers, incomplete punctuation, conversational speech, and ASR errors.

The official wtpsplit project provides language/domain LoRAs for `sat-3l` and `sat-12l`, including TED, UD, OPUS100, and Ersatz adapters. Use `ted2020-corrupted` as the primary treatment.

## Models to Compare

- [ ] M1: `sat-1l` base — existing baseline
- [ ] M2: `sat-3l` base
- [ ] M3: `sat-3l + ted2020-corrupted LoRA`
- [ ] M4: `sat-3l + ud LoRA` — control for separating language adaptation from ASR-domain adaptation

Languages:

- [ ] `en`
- [ ] `ko`
- [ ] `ja`
- [ ] `zh`

> Do not assume a `sat-3l` LoRA is compatible with `sat-1l` or `sat-3l-sm`. Use the exact matching base variant.

## Input Data

Replay the existing Experiment 10/11 candidates and snapshots without making new API calls.

- Include every candidate regardless of whether it originally triggered.
- Labels: `WAIT`, `PROVISIONAL_TRANSLATE`, `SAFE_APPEND`
- Reuse the same raw/confirmed source, future T4, and FULL continuation.
- Split by **turn**, not by candidate, to avoid leakage.

## Evaluation

Do not reuse the old `0.25` threshold. LoRA score scales may differ, so calibrate each model separately.

- [ ] PR-AUC
- [ ] ROC-AUC — secondary metric
- [ ] Candidate recall / turn coverage at precision ≥ 85%
- [ ] Precision at turn coverage 50% / 65% / 80%
- [ ] Score distributions for useful candidates vs. `WAIT`
- [ ] Top-1 cut ranking accuracy within each window
- [ ] Margin between the best positive cut and the second-best cut
- [ ] Per-language results
- [ ] Per-model p50 / p90 inference latency
- [ ] Decision flips / source rollback

## LoRA Promotion Criteria

Promote M3 as the default SaT backend for the later Qwen and Soniox experiments if it satisfies at least one of the following:

- [ ] At precision ≥ 85%, turn coverage improves by **at least +10 percentage points** over M2.
- [ ] PR-AUC improves meaningfully over M2.
- [ ] Top-1 cut accuracy improves by **at least +10 percentage points**.
- [ ] At equal coverage, calls/turn or false triggers decrease by **at least 20%**.

If there is no improvement, continue with whichever of `sat-3l base` or `sat-1l base` performs better. Do not cancel the structural experiments solely because the LoRA fails.

## ONNX Validation

- [ ] Reproduce PyTorch LoRA results.
- [ ] If promoted, export a LoRA-merged ONNX model with `export_to_onnx_sat.py`.
- [ ] Verify CPU inference latency and output equivalence.

---

# P1 — Qwen Confirmed-Prefix Internal Boundary Selection

## Problem

Qwen's confirmed `text` is rollback-free, but it grows in very small increments. The policy `end of stableText = candidate boundary` produces many locations that are not meaningful translation boundaries and therefore has poor recall.

## Q1. Fixed-Lag Internal Segmentation

Instead of treating the current end of `text` as the boundary, re-segment the **confirmed prefix** after the last committed cut.

```text
confirmed text:  [Sentence A.][Partial sentence B continues]
                              ^ internal cut candidate
future context:                                      [text/stash T4]
```

Requirements:

- [ ] The entire source before the cut must be inside confirmed `text`.
- [ ] At least future T4 must be available after the cut using the exact SaT tokenizer.
- [ ] Exclude the trailing T4 region from source-cut candidates.
- [ ] Exclude positions before the last committed cut.
- [ ] Select only one top-ranked cut per evaluation.

## Q2. Viterbi / Length Constraints Instead of Only Thresholding

Compare:

- [ ] Simple `score > threshold`
- [ ] SaT raw probability + minimum / target / maximum length prior
- [ ] Greedy top-1 local maximum
- [ ] Viterbi global-best cut

Initial sweep:

- Minimum length: 4 / 8 SaT tokens
- Target length: 12 / 18 / 24 tokens
- Maximum length: 24 / 32 / 40 tokens

## Q3. Boundary Persistence

Measure whether the same SaT peak remains stable as future context grows, rather than measuring whether the source string itself remains unchanged.

- [ ] The same normalized cut remains top-1 or top-2 across 1 / 2 / 3 revisions.
- [ ] The cut score increases or keeps a minimum margin.
- [ ] Compare added latency against precision improvement.

## Q4. LocalAgreement Pseudo-Confirmed Stash

Test whether a safe region can be created slightly ahead of native Qwen `text`.

- [ ] Common prefix across 2 / 3 consecutive `text + stash` snapshots
- [ ] Compare raw matching with normalization of whitespace, punctuation, and case
- [ ] Measure time gained over native `text`
- [ ] Measure pseudo-confirmed rollback and semantic change rate
- [ ] Run internal SaT cut selection inside the pseudo-confirmed region

## Q5. Bounded-Latency Trigger

Do not let SaT alone decide when to run.

- [ ] 750 / 1000 / 1250 ms since the last preview
- [ ] 8 / 12 / 16 new SaT tokens
- [ ] Maximum wait of 2000 / 2500 / 3000 ms
- [ ] When triggered, select only the top-1 SaT cut inside the safe window

## Q6. Qwen Checkpoint Comparison

When possible, compare the stable alias with the latest explicitly versioned realtime checkpoint using identical audio.

Measure:

- Time to first visible text / first confirmed text
- `text` growth granularity
- `stash` rollback
- Early short hallucinations
- Punctuation density
- Internal SaT cut precision / coverage

## Qwen Success Criteria

- [ ] Precision ≥ 85%
- [ ] Reference-positive turn coverage ≥ 65%
- [ ] Preview p50 ≤ 3.0 s
- [ ] Preview p90 ≤ 4.5 s
- [ ] Calls/turn ≤ 1.2
- [ ] Semantic rollback ≤ 5%
- [ ] Early-hallucination triggers ≤ 2%

---

# P2 — Soniox Local VAD + Manual Finalize + SaT

## Assumption

Do not use server-side VAD or semantic endpointing. Keep the existing product design in which **local VAD detects a pause and sends manual finalize to Soniox**.

Keep these concepts separate:

```text
ASR finalize = source tokens will no longer change
SaT approve   = the source is a suitable semantic unit to translate now
```

If manual finalize occurs but SaT returns `WAIT`, append the next finalized chunk to the application's logical source buffer.

## Prerequisite Fixes

- [ ] Prevent the Soniox adapter from dropping the `<fin>` control token when duplicate or out-of-order final tokens appear in the same event.
- [ ] Process every raw event to completion and handle control tokens independently.
- [ ] Add a `stableText` monotonicity test.
- [ ] Record whether one `<fin>` is received for every manual-finalize request.

## S1. VAD-Finalized Candidate + Raw T4 + SaT

```text
local VAD pause
  → manual finalize
  → immutable final source candidate
  → next raw T4
  → SaT
      approve: MT preview
      WAIT: logically merge with the next finalized chunk
```

VAD sweep:

- [ ] Trailing silence: 200 / 300 / 400 / 500 ms
- [ ] Minimum speech duration: 1.0 / 1.5 / 2.0 s
- [ ] Minimum finalize interval: 500 / 750 / 1000 ms
- [ ] Minimum new SaT tokens before finalize: 2 / 4 / 8

Measure:

- Finalize-candidate precision / recall / turn coverage
- Semantic speech boundary → `<fin>` latency
- `<fin>` → raw T4 latency
- Preview p50 / p90
- ASR error and punctuation changes
- Finalize/turn and MT calls/turn
- Final logical source accuracy after `WAIT` merges

## S2. LocalAgreement Soft-Stable Frontier

Create a pseudo-confirmed source earlier than native finalization.

- [ ] Common prefix across 2 / 3 consecutive non-final token snapshots
- [ ] Token-timestamp alignment
- [ ] Whether punctuation-only changes should be ignored
- [ ] Handling of subword/tokenization changes
- [ ] Source rollback and semantic change rate

## S3. Token Age + Confidence

Use Soniox token `end_ms`, the processed-audio frontier, and confidence to define soft-stable tokens.

Initial sweep:

- Token age: 300 / 500 / 750 / 1000 ms
- Agreement count: 2 / 3 snapshots
- Confidence: 0.75 / 0.85 / 0.95

```text
softStable(token) =
  token_age >= A
  AND agreement_count >= N
  AND confidence >= C
```

Select the top-1 SaT cut inside the soft-stable window.

## S4. VAD-Final + Soft-Stable Hybrid

- [ ] Use LocalAgreement / age-based frontier during continuous speech.
- [ ] Use manual finalize when local VAD detects a pause.
- [ ] If SaT finds no acceptable cut, return `WAIT` and merge with the next chunk.
- [ ] If the maximum wait is exceeded, select the top-ranked cut inside the finalized window or use endpoint fallback.

## S5. Bounded-Latency Viterbi

- [ ] Minimum / target / maximum source length
- [ ] Time budget since the previous preview
- [ ] Select only the top-1 cut inside the safe window
- [ ] Use a hard maximum wait so coverage does not collapse solely because no threshold is met

## Soniox Success Criteria

- [ ] Precision ≥ 85%
- [ ] Reference-positive turn coverage ≥ 65%
- [ ] Preview p50 ≤ 3.5 s
- [ ] Preview p90 ≤ 5.0 s
- [ ] Calls/turn ≤ 1.2
- [ ] Semantic rollback ≤ 5%
- [ ] Finalize/turn remains within the product's acceptable range
- [ ] ASR quality degradation versus the current finalization policy remains acceptable

---

# P3 — Shared Follow-Up Experiments

Run these only if P1/P2 fail the precision or coverage requirements.

## C1. SaT Top-k + MT-Aware Reranking

- [ ] Send only SaT top-2 / top-3 cuts to MT.
- [ ] Measure target-prefix agreement between `T(source@cut)` and `T(source@cut + future)`.
- [ ] Distinguish open/incomplete translations from complete translations.
- [ ] Apply only in ambiguous SaT-score regions.
- [ ] Use batching and caching to control call cost.

## C2. Speculative MT

- [ ] Start MT on the top raw / soft-stable candidate.
- [ ] Do not display the result before the source is confirmed.
- [ ] Show the cached translation immediately when the candidate becomes confirmed.
- [ ] Discard the result if the candidate disappears.

## C3. Rolling Retranslation + SaT Commit

Fallback when no reliable internal cut can be found first.

- [ ] Retranslate the latest partial window at a limited cadence.
- [ ] Display the result as a replaceable preview.
- [ ] Use SaT for source commit cuts and window trimming.
- [ ] Re-translate and commit at endpoint / hard final.

---

# Data and Labeling Rules

## Reuse Existing Logs

First replay the existing Experiment 10/11 logs for P0, Qwen internal segmentation, LocalAgreement, boundary persistence, and Soniox token-age experiments.

## New Captures Required For

- Qwen checkpoint comparison
- Soniox VAD silence / finalize parameter sweep
- Measuring the effect of actual manual finalization on ASR quality

## Validation Dataset

Use the existing 19 turns only for exploration and calibration. Use a separate dataset for final validation.

Recommended minimum:

- 20 English turns
- 20 Korean turns
- 20 Japanese turns
- 20 Chinese turns
- 80 turns total
- Each turn: 8–25 seconds of natural multi-clause speech

Include:

- Condition-result constructions
- Long clauses connected by conjunctions
- Fillers and repetition
- Self-correction
- Enumeration
- Multiple short complete sentences in sequence
- Long speech without pauses
- Speech with frequent micro-pauses

## Candidate Labels

Store and label every candidate regardless of whether it triggers.

- `WAIT`
- `PROVISIONAL_TRANSLATE`
- `SAFE_APPEND`

Evaluation mapping:

- Scheduling positive = `PROVISIONAL_TRANSLATE` + `SAFE_APPEND`
- Permanent positive = `SAFE_APPEND`

---

# Shared Log Schema

```text
provider
language
turn_id
revision
candidate_id
source_policy
source_region
candidate_text
candidate_cut
future_t4
full_continuation
candidate_audio_end_ms
candidate_created_at_ms
source_confirmed_at_ms
t4_available_at_ms
sat_model
sat_adapter
sat_probability
sat_rank
sat_margin
persistence_count
agreement_count
token_age_ms
mean_confidence
vad_silence_ms
manual_finalize_sent_at_ms
fin_received_at_ms
mt_started_at_ms
mt_completed_at_ms
later_source_changed
semantic_rollback
label
```

---

# Expected Outputs

```text
experiments/experiment-12-sat-recovery/
├─ configs/
│  ├─ lora_models.yaml
│  ├─ qwen_profiles.yaml
│  └─ soniox_profiles.yaml
├─ replay/
├─ raw/
├─ normalized/
├─ candidates.parquet
├─ labels.csv
├─ lora_results.csv
├─ qwen_results.csv
├─ soniox_results.csv
├─ latency_summary.json
└─ experiment_report.md
```

> This path is not a requirement to create a new structure. If the local experimental branch already has established directories, file naming, or CLI conventions, reuse them and add only the required outputs to the existing structure.

---

# Execution Order and Stop Conditions

1. [ ] **P0 LoRA re-evaluation**
2. [ ] Select and freeze the promoted SaT backend
3. [ ] Replay Qwen internal fixed-lag + Viterbi + persistence
4. [ ] Fix Soniox adapter `<fin>` handling
5. [ ] Replay Soniox LocalAgreement + token age
6. [ ] Capture new Soniox local-VAD/manual-finalize runs
7. [ ] Compare Qwen/Soniox bounded-latency combinations
8. [ ] Validate on a new independent 80-turn dataset
9. [ ] Run MT-aware / rolling fallback experiments only if needed

Stop conditions:

- Do not repeatedly micro-tune only threshold, cooldown, or length rules on the same data.
- Stop a standalone strategy if turn coverage remains below 50% at precision ≥ 85%.
- Treat the result as overfitting if independent-validation coverage drops by more than 15 percentage points from exploration.
- If calls/turn remains above 2, classify the approach as rolling/speculative rather than as a standalone trigger strategy.

---

# Product Integration Target

```text
NormalizedAsrSnapshot
  → provider-specific stable-region builder
  → SaT verifier / Viterbi cut selector
  → replaceable translation preview
  → endpoint / hard final
  → committed translation
```

Provider profiles:

```text
Deepgram:
  native confirmed + raw T4 + SaT

Qwen:
  confirmed prefix / LocalAgreement window
  + internal SaT cut selection

Soniox:
  local-VAD manual-finalized source
  or agreement/age-based soft-stable source
  + SaT cut selection
```

This experiment does not target permanent append behavior. Even an approved boundary should initially produce only a replaceable preview, followed by final retranslation and commit at endpoint / hard final.

---

## References

- wtpsplit / SaT: https://github.com/segment-any-text/wtpsplit
- `sat-3l` LoRA directory: https://huggingface.co/segment-any-text/sat-3l/tree/main/loras
- `ted2020-corrupted`: the closest official adapter candidate for ASR-style transcribed speech
